File: integrations/java/src/main/java/io/contextpruner/sidecar/ContextPrunerSidecarClient.java
package io.contextpruner.sidecar;

import java.io.IOException;
import java.net.URI;
import java.net.URLEncoder;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.nio.charset.StandardCharsets;
import java.time.Duration;
import java.util.Collections;
import java.util.LinkedHashMap;
import java.util.Map;

/**
 * Zero-dependency Java 11+ client for the Context-Pruner lifecycle Sidecar.
 *
 * <p>The host owns the authoritative full history. {@link #beforeModel} returns
 * a temporary model view and must never replace that full history.</p>
 */
public final class ContextPrunerSidecarClient implements AutoCloseable {
    private final URI baseUri;
    private final String authToken;
    private final Duration timeout;
    private final HttpClient http;

    public ContextPrunerSidecarClient(String baseUrl, String authToken, Duration timeout) {
        if (baseUrl == null || baseUrl.trim().isEmpty()) {
            throw new IllegalArgumentException("baseUrl is required");
        }
        URI parsed = URI.create(baseUrl.trim());
        if (!("http".equalsIgnoreCase(parsed.getScheme()) || "https".equalsIgnoreCase(parsed.getScheme()))) {
            throw new IllegalArgumentException("baseUrl must use http or https");
        }
        String normalized = parsed.toString();
        this.baseUri = URI.create(normalized.endsWith("/") ? normalized.substring(0, normalized.length() - 1) : normalized);
        this.authToken = authToken == null ? "" : authToken;
        this.timeout = timeout == null ? Duration.ofSeconds(30) : timeout;
        if (this.timeout.isZero() || this.timeout.isNegative()) {
            throw new IllegalArgumentException("timeout must be positive");
        }
        this.http = HttpClient.newBuilder().connectTimeout(this.timeout).build();
    }

    public ContextPrunerSidecarClient(String baseUrl, String authToken) {
        this(baseUrl, authToken, Duration.ofSeconds(30));
    }

    public Map<String, Object> health() throws IOException, InterruptedException {
        return request("GET", "/health", null, false);
    }

    public Map<String, Object> capabilities() throws IOException, InterruptedException {
        return request("GET", "/v1/capabilities", null, true);
    }

    public Map<String, Object> beforeModel(String sessionId, Map<String, Object> payload)
            throws IOException, InterruptedException {
        return sessionRequest(sessionId, "before-model", "POST", payload);
    }

    public Map<String, Object> afterModel(String sessionId, Object output)
            throws IOException, InterruptedException {
        return sessionRequest(sessionId, "after-model", "POST", map("output", output));
    }

    public Map<String, Object> afterTool(String sessionId, Object output)
            throws IOException, InterruptedException {
        return sessionRequest(sessionId, "after-tool", "POST", map("output", output));
    }

    public Map<String, Object> onError(String sessionId, Object error, String query, boolean recover)
            throws IOException, InterruptedException {
        Map<String, Object> payload = new LinkedHashMap<>();
        payload.put("error", String.valueOf(error));
        payload.put("query", query == null ? "" : query);
        payload.put("recover", recover);
        return sessionRequest(sessionId, "error", "POST", payload);
    }

    public Map<String, Object> finalizeSession(String sessionId, Map<String, Object> metadata)
            throws IOException, InterruptedException {
        return sessionRequest(sessionId, "finalize", "POST", map("metadata", metadata == null ? Collections.emptyMap() : metadata));
    }

    public Map<String, Object> getState(String sessionId) throws IOException, InterruptedException {
        return sessionRequest(sessionId, "state", "GET", null);
    }

    public Map<String, Object> restoreState(
            String sessionId,
            Map<String, Object> lifecycleState,
            String taskState,
            Map<String, Object> config
    ) throws IOException, InterruptedException {
        Map<String, Object> payload = new LinkedHashMap<>();
        if (config != null) payload.putAll(config);
        payload.put("lifecycle_state", lifecycleState);
        payload.put("task_state", taskState == null ? "" : taskState);
        return sessionRequest(sessionId, "state", "PUT", payload);
    }

    public Map<String, Object> deleteSession(String sessionId) throws IOException, InterruptedException {
        return sessionRequest(sessionId, "", "DELETE", null);
    }

    private Map<String, Object> sessionRequest(
            String sessionId, String suffix, String method, Map<String, Object> payload
    ) throws IOException, InterruptedException {
        if (sessionId == null || sessionId.trim().isEmpty()) {
            throw new IllegalArgumentException("sessionId is required");
        }
        String encoded = URLEncoder.encode(sessionId.trim(), StandardCharsets.UTF_8).replace("+", "%20");
        String tail = suffix == null || suffix.isEmpty() ? "" : "/" + suffix;
        return request(method, "/v1/sessions/" + encoded + tail, payload, true);
    }

    private Map<String, Object> request(
            String method, String path, Map<String, Object> payload, boolean authenticated
    ) throws IOException, InterruptedException {
        HttpRequest.Builder builder = HttpRequest.newBuilder(baseUri.resolve(path))
                .timeout(timeout)
                .header("Accept", "application/json");
        if (authenticated && !authToken.isEmpty()) {
            builder.header("Authorization", "Bearer " + authToken);
        }
        if (payload == null) {
            builder.method(method, HttpRequest.BodyPublishers.noBody());
        } else {
            builder.header("Content-Type", "application/json")
                    .method(method, HttpRequest.BodyPublishers.ofString(JsonCodec.stringify(payload), StandardCharsets.UTF_8));
        }
        HttpResponse<String> response = http.send(
                builder.build(), HttpResponse.BodyHandlers.ofString(StandardCharsets.UTF_8)
        );
        Map<String, Object> data;
        try {
            data = response.body() == null || response.body().isEmpty()
                    ? new LinkedHashMap<>()
                    : JsonCodec.parseObject(response.body());
        } catch (IllegalArgumentException error) {
            throw new SidecarException("Sidecar returned invalid JSON", response.statusCode(), "invalid_json", response.body());
        }
        if (response.statusCode() < 200 || response.statusCode() >= 300) {
            Object errorValue = data.get("error");
            Map<?, ?> remote = errorValue instanceof Map ? (Map<?, ?>) errorValue : Collections.emptyMap();
            Object remoteMessage = remote.containsKey("message")
                    ? remote.get("message")
                    : "Sidecar request failed with HTTP " + response.statusCode();
            Object remoteCode = remote.containsKey("code") ? remote.get("code") : "";
            String message = String.valueOf(remoteMessage);
            String code = String.valueOf(remoteCode);
            throw new SidecarException(message, response.statusCode(), code, data);
        }
        return data;
    }

    private static Map<String, Object> map(String key, Object value) {
        Map<String, Object> result = new LinkedHashMap<>();
        result.put(key, value);
        return result;
    }

    @Override
    public void close() {
        // java.net.http.HttpClient does not own a closeable resource in Java 11.
    }

    public static final class SidecarException extends IOException {
        private static final long serialVersionUID = 1L;
        private final int status;
        private final String code;
        private final Object details;

        public SidecarException(String message, int status, String code, Object details) {
            super(message);
            this.status = status;
            this.code = code == null ? "" : code;
            this.details = details;
        }

        public int getStatus() { return status; }
        public String getCode() { return code; }
        public Object getDetails() { return details; }
    }
}

File: integrations/javascript/context-pruner-sidecar-client.mjs
/**
 * Zero-dependency JavaScript client for the Context-Pruner lifecycle Sidecar.
 *
 * The host remains the owner of the full conversation. `beforeModel()` returns
 * a temporary model view and must never replace the host's authoritative history.
 */

export class ContextPrunerSidecarError extends Error {
  constructor(message, { status = 0, code = '', details = null } = {}) {
    super(message);
    this.name = 'ContextPrunerSidecarError';
    this.status = status;
    this.code = code;
    this.details = details;
  }
}

export class ContextPrunerSidecarClient {
  constructor({ baseUrl, authToken = '', timeoutMs = 30_000, fetchImpl = globalThis.fetch } = {}) {
    if (typeof fetchImpl !== 'function') {
      throw new TypeError('A WHATWG-compatible fetch implementation is required');
    }
    const parsed = new URL(String(baseUrl || ''));
    if (!['http:', 'https:'].includes(parsed.protocol)) {
      throw new TypeError('baseUrl must use http or https');
    }
    this.baseUrl = parsed.toString().replace(/\/$/, '');
    this.authToken = String(authToken || '');
    this.timeoutMs = Math.max(1, Number(timeoutMs) || 30_000);
    this.fetchImpl = fetchImpl;
  }

  async health() {
    return this.#request('GET', '/health', undefined, false);
  }

  async capabilities() {
    return this.#request('GET', '/v1/capabilities', undefined, true);
  }

  async beforeModel(sessionId, payload) {
    return this.#sessionRequest(sessionId, 'before-model', 'POST', payload);
  }

  async afterModel(sessionId, output) {
    return this.#sessionRequest(sessionId, 'after-model', 'POST', { output });
  }

  async afterTool(sessionId, output) {
    return this.#sessionRequest(sessionId, 'after-tool', 'POST', { output });
  }

  async onError(sessionId, error, { query = '', recover = true } = {}) {
    return this.#sessionRequest(sessionId, 'error', 'POST', {
      error: String(error),
      query: String(query),
      recover: Boolean(recover),
    });
  }

  async finalize(sessionId, metadata = {}) {
    return this.#sessionRequest(sessionId, 'finalize', 'POST', { metadata });
  }

  async getState(sessionId) {
    return this.#sessionRequest(sessionId, 'state', 'GET');
  }

  async restoreState(sessionId, lifecycleState, { taskState = '', ...config } = {}) {
    return this.#sessionRequest(sessionId, 'state', 'PUT', {
      lifecycle_state: lifecycleState,
      task_state: String(taskState),
      ...config,
    });
  }

  async deleteSession(sessionId) {
    return this.#sessionRequest(sessionId, '', 'DELETE');
  }

  async #sessionRequest(sessionId, suffix, method, payload) {
    const normalized = String(sessionId || '').trim();
    if (!normalized) {
      throw new TypeError('sessionId is required');
    }
    const tail = suffix ? `/${suffix}` : '';
    return this.#request(
      method,
      `/v1/sessions/${encodeURIComponent(normalized)}${tail}`,
      payload,
      true,
    );
  }

  async #request(method, path, payload, authenticated) {
    const controller = new AbortController();
    const timeout = setTimeout(() => controller.abort(), this.timeoutMs);
    const headers = { Accept: 'application/json' };
    if (payload !== undefined) {
      headers['Content-Type'] = 'application/json';
    }
    if (authenticated && this.authToken) {
      headers.Authorization = `Bearer ${this.authToken}`;
    }
    try {
      const response = await this.fetchImpl(`${this.baseUrl}${path}`, {
        method,
        headers,
        body: payload === undefined ? undefined : JSON.stringify(payload),
        signal: controller.signal,
      });
      const text = await response.text();
      let data = {};
      if (text) {
        try {
          data = JSON.parse(text);
        } catch (error) {
          throw new ContextPrunerSidecarError('Sidecar returned invalid JSON', {
            status: response.status,
            details: text.slice(0, 300),
          });
        }
      }
      if (!response.ok) {
        const remote = data?.error || {};
        throw new ContextPrunerSidecarError(
          String(remote.message || `Sidecar request failed with HTTP ${response.status}`),
          {
            status: response.status,
            code: String(remote.code || ''),
            details: data,
          },
        );
      }
      return data;
    } catch (error) {
      if (error instanceof ContextPrunerSidecarError) {
        throw error;
      }
      const timedOut = error?.name === 'AbortError';
      throw new ContextPrunerSidecarError(
        timedOut ? `Sidecar request timed out after ${this.timeoutMs} ms` : String(error?.message || error),
        { code: timedOut ? 'timeout' : 'transport_error' },
      );
    } finally {
      clearTimeout(timeout);
    }
  }
}
